### PR TITLE
Append Doxygen-friendly comments

### LINKED ISSUE #8: Add doxygen friendly comments
We haven't written any comment for doxygen at this moment. We should implement the comments for all methods

### PR DESCRIPTION
**Pull request by splasky**
Function descriptions was missing in several files.
Add these descriptions for doxygen.

Closes #8

### CODE REVIEW

**Review by jserv — CHANGES_REQUESTED** [review]
We distinguish from each other:
1. Documentation: general descriptions for high-level concepts, build/test/validation instructions, technical reports, etc.
2. In-program comments: used with Doxygen for something similar;

File: utils/crypto_utils.h [hunk 1]
```diff
@@ -20,16 +20,90 @@ extern "C" {
...
+ * - Ciphertext text new length on success
+ * - 0 on error
+ */
 int encrypt(const char *plaintext, int plaintext_len, char *ciphertext, int ciphertext_len, uint8_t iv[AES_BLOCK_SIZE],
             uint8_t key[AES_CBC_KEY_SIZE], uint8_t device_id[IMSI_LEN + 1]);
-int decrypt(const char *ciphertext, int ciphertext_len, char *plaintext, int plaintext_len, uint8_t iv[AES_BLOCK_SIZE],
-            uint8_t key[AES_CBC_KEY_SIZE]);
+
+/**
+ * @brief Decrypt ciphertext to plaintext
```
Comment (jserv) [inline utils/crypto_utils.h]: Simply say, "Decrypt ciphertext"

Comment (jserv) [reply utils/crypto_utils.h]: It is not necessary to mention "to plaintext" since "decrypt" always implies that.

File: utils/crypto_utils.h [hunk 2]
```diff
@@ -20,16 +20,90 @@ extern "C" {
...
+ * @param[in] iv[AES_BLOCK_SIZE] Initialization vector
+ * @param[in] key[AES_CBC_KEY_SIZE] Decryption key
+ *
+ * @return #retcode_t
+ */
+retcode_t decrypt(const char *ciphertext, int ciphertext_len, char *plaintext, int plaintext_len,
+                  uint8_t iv[AES_BLOCK_SIZE], uint8_t key[AES_CBC_KEY_SIZE]);
+
+/**
+ * @brief Implementation of aes encrypt
```
Comment (jserv) [inline utils/crypto_utils.h]: Always use uppercase "AES." We would say, "AES encryption algorithm."

File: connectivity/conn_http.h [hunk 3]
```diff
@@ -23,31 +23,98 @@ extern "C" {
...
 
+/**
+ * @brief Open http connection
+ *
+ * @param[in, out] info Context for http connection
+ * @param[in] seed_nonce Personalization data, that is device-specific identifiers. Can be NULL.
+ * @param[in] host Http host to connect
+ * @param[in] port Http port to connect
+ *
+ * @return #retcode_t
```
Comment (howjmay) [inline connectivity/conn_http.h]: Follow the pattern in line 115

File: connectivity/conn_http.h [hunk 4]
```diff
@@ -23,31 +23,98 @@ extern "C" {
...
 retcode_t http_open(connect_info_t *const info, char const *const seed_nonce, char const *const host,
                     char const *const port);
+
+/**
+ * @brief Send request to http connection
+ *
+ * @param[in] info Context for http connection
+ * @param[in] req  Buffer holding the data
+ *
+ * @return #retcode_t
```
Comment (howjmay) [inline connectivity/conn_http.h]: ditto

File: connectivity/conn_http.h [hunk 5]
```diff
@@ -23,31 +23,98 @@ extern "C" {
...
 retcode_t http_send_request(connect_info_t *const info, const char *req);
+
+/**
+ * @brief Read response from http server
+ *
+ * @param[in] info Context for http connection
+ * @param[out] res Buffer that will hold the data
+ * @param[out] res_len Length of res
+ *
+ * @return #retcode_t
```
Comment (howjmay) [inline connectivity/conn_http.h]: ditto

File: connectivity/conn_http.h [hunk 6]
```diff
@@ -23,31 +23,98 @@ extern "C" {
...
+ */
 retcode_t http_read_response(connect_info_t *const info, char *res, size_t res_len);
+
+/**
+ * @brief Close http connection
+ *
+ * @param[in] info Context for http connection
+ *
+ * @return #retcode_t
```
Comment (howjmay) [inline connectivity/conn_http.h]: ditto

File: connectivity/conn_http.h [hunk 7]
```diff
@@ -23,31 +23,98 @@ extern "C" {
...
+/**
+ * @brief Set post request message
+ *
+ * @param[in] api API for post request to https server, example: transaction/. Must be string.
+ * @param[in] host Http host to connect
+ * @param[in] port Http port to connect
+ * @param[in] req_body Pointer to post request body
+ * @param[out] out Post request message
+ *
+ * @return #retcode_t
```
Comment (howjmay) [inline connectivity/conn_http.h]: ditto

File: connectivity/conn_http.h [hunk 8]
```diff
@@ -23,31 +23,98 @@ extern "C" {
...
                            char const *const req_body, char **out);
+/**
+ * @brief Set get request message
+ *
+ * @param[in] api API for post request to https server, example: transaction/. Must be string.
+ * @param[in] host Http host to connect
+ * @param[in] port Http port to connect
+ * @param[out] out Get request message
+ *
+ * @return #retcode_t
```
Comment (howjmay) [inline connectivity/conn_http.h]: ditto

File: utils/crypto_utils.h [hunk 9]
```diff
@@ -20,16 +20,90 @@ extern "C" {
...
+ * @return #retcode_t
+ */
+retcode_t get_device_id(const char *device_id);
+
+/**
+ * @brief Gets aes key from hashchain
+ *
+ * @param[out] key Device key output
+ *
+ * @return #retcode_t
```
Comment (howjmay) [inline utils/crypto_utils.h]: ditto

File: utils/crypto_utils.h [hunk 10]
```diff
@@ -20,16 +20,90 @@ extern "C" {
...
+ */
+retcode_t get_aes_key(const uint8_t *key);
+
+/**
+ * @brief Encrypt plaintext to ciphertext
+ *
+ * @param[in] plaintext The text to be encrypted
+ * @param[in] plaintext_len Plaintext length
+ * @param[out] ciphertext Ciphrtext
```
Comment (howjmay) [inline utils/crypto_utils.h:50]: Move output parameter in the end of the parameters.

File: utils/crypto_utils.h [hunk 11]
```diff
@@ -20,16 +20,90 @@ extern "C" {
...
             uint8_t key[AES_CBC_KEY_SIZE], uint8_t device_id[IMSI_LEN + 1]);
-int decrypt(const char *ciphertext, int ciphertext_len, char *plaintext, int plaintext_len, uint8_t iv[AES_BLOCK_SIZE],
-            uint8_t key[AES_CBC_KEY_SIZE]);
+
+/**
+ * @brief Decrypt ciphertext to plaintext
+ *
+ * @param[in] ciphertext Ciphertext
+ * @param[in] ciphertext_len Ciphertext length
+ * @param[out] plaintext Plaintext
```
Comment (howjmay) [inline utils/crypto_utils.h:68]: ditto

File: utils/crypto_utils.h [hunk 12]
```diff
@@ -20,16 +20,90 @@ extern "C" {
...
+
+/**
+ * @brief Implementation of aes encrypt
+ *
+ * @param[in] plaintext Plaintext
+ * @param[in] plaintext_len Plaintext length
+ * @param[in] key Encryption key
+ * @param[in] keybits Length of key
+ * @param[in, out] iv[AES_BLOCK_SIZE] Initialization vector
+ * @param[out] ciphertext Ciphertext
```
Comment (howjmay) [inline utils/crypto_utils.h:86]: ditto

File: utils/crypto_utils.h [hunk 13]
```diff
@@ -20,16 +20,90 @@ extern "C" {
...
+
+/**
+ * @brief Implementation of aes decrypt
+ *
+ * @param[in] ciphertext Ciphertext
+ * @param[in] ciphertext_len Ciphertext length
+ * @param[in] key Decryption key
+ * @param[in] keybits Length of decryption key
+ * @param[in] iv[AES_BLOCK_SIZE] Initialization vector
+ * @param[out] plaintext Plaintext
```
Comment (howjmay) [inline utils/crypto_utils.h:104]: ditto

File: utils/serializer.h [hunk 14]
```diff
@@ -15,8 +15,37 @@
...
+ * @param[in] ciphertext_len Length of ciphertext
+ * @param[in] ciphertext Ciphertext to be serialized
+ * @param[out] out_msg Pointer to output message
+ * @param[out] out_msg_len Pointer to length of serialized message
+ *
+ * @return
+ * - 0 on success
+ * - non-zero on error
+ */
 int serialize_msg(const uint8_t *iv, uint32_t ciphertext_len, const char *ciphertext, char *out_msg,
```
Comment (howjmay) [inline utils/serializer.h:35]: How about using `retcode_t`

File: utils/serializer.h [hunk 15]
```diff
@@ -15,8 +15,37 @@
...
+ * @param[in] iv Pointer to initialize vector
+ * @param[out] ciphertext_len Pointer to plaintext length
+ * @param[out] ciphertext Pointer to plaintext output array
+ *
+ * @return
+ * - 0 on success
+ * - non-zero on error
+ * @see #serialize_msg
+ */
 int deserialize_msg(char *msg, const uint8_t *iv, uint32_t *ciphertext_len, char *ciphertext);
```
Comment (howjmay) [inline utils/serializer.h:51]: ditto

**Review by howjmay — CHANGES_REQUESTED** [review]

File: utils/crypto_utils.h (hunk 10 above)
Comment (splasky) [reply utils/crypto_utils.h:50]: Already mark this at #54.

File: utils/crypto_utils.h (hunk 11 above)
Comment (splasky) [reply utils/crypto_utils.h:68]: Already mark this at #54.

File: utils/crypto_utils.h (hunk 13 above)
Comment (splasky) [reply utils/crypto_utils.h:104]: Already mark this at #54.

File: utils/crypto_utils.h (hunk 12 above)
Comment (splasky) [reply utils/crypto_utils.h:86]: Already mark this at #54.

File: utils/serializer.h (hunk 14 above)
Comment (splasky) [reply utils/serializer.h:35]: Mark at #61 .

File: utils/serializer.h (hunk 15 above)
Comment (splasky) [reply utils/serializer.h:51]: Mark at #61 .

File: utils/crypto_utils.h [hunk 16]
```diff
@@ -20,16 +20,90 @@ extern "C" {
...
 #define IMSI_LEN 15
 
-int get_device_id(const char *device_id);
-int get_aes_key(const uint8_t *key);
+/**
+ * @brief Gets device IMSI or physical ID
+ *
+ * @param[out] device_id Device id output
+ *
+ * @return #retcode_t
```
Comment (howjmay) [inline utils/crypto_utils.h]: ditto

**Review by howjmay — CHANGES_REQUESTED** [review]

File: utils/crypto_utils.h [hunk 17]
```diff
@@ -20,16 +20,94 @@ extern "C" {
 #define MAXLINE 1024
 #define IMSI_LEN 15
 
+/**
+ * @brief Gets device IMSI or physical ID
```
Comment (jserv) [inline utils/crypto_utils.h]: Use "Get" or "Obtain" for consistency.

File: utils/serializer.h [hunk 18]
```diff
@@ -15,8 +15,37 @@
 extern "C" {
 #endif
 
+/**
+ * @brief Serializing ciphertext and initialize vector together, the out put message
```
Comment (jserv) [inline utils/serializer.h]: Should be "Serialize"

File: utils/serializer.h [hunk 19]
```diff
@@ -15,8 +15,37 @@
...
+ * format show as below:
+ * | initialize vector(16 bytes) | ciphertext length(10 bytes) | ciphertext |
```
Comment (jserv) [inline utils/serializer.h]: Improve the description with pretty formatting.

File: connectivity/conn_http.h [hunk 20]
```diff
@@ -23,31 +23,121 @@ extern "C" {
 #include "mbedtls/net.h"
 #include "mbedtls/ssl.h"
 
-#define HTTP_OK 200
-
+/**< Global pointer points to http resource body */
```
Comment (jserv) [inline connectivity/conn_http.h]: "Global pointer points to http resource body" What?
Improve it.

Comment (splasky) [reply connectivity/conn_http.h]: I know this is the bad coding style. This global variable has existed for a long time. For this subject, I will just update the doxygen comment and mark this for a new issue.

Comment (splasky) [reply connectivity/conn_http.h]: Mark at #62 .

File: connectivity/conn_http.h [hunk 21]
```diff
@@ -23,31 +23,123 @@ extern "C" {
 #include "mbedtls/net.h"
 #include "mbedtls/ssl.h"
 
-#define HTTP_OK 200
-
+/**< Global pointer points to http resource body for passing response body to http_parser.
+ * This will be removed or change to static pointer as soon as possible. Don't use this
```
Comment (howjmay) [inline connectivity/conn_http.h]: This variable will be removed or changed

File: connectivity/conn_http.h [hunk 22]
```diff
@@ -23,31 +23,123 @@ extern "C" {
...
+ * - non-zero on error
+ * @see #retcode_t
+ */
 retcode_t http_send_request(connect_info_t *const info, const char *req);
+
+/**
+ * @brief Read response from http server
+ *
+ * @param[in] info Context for http connection
+ * @param[out] res Buffer that will hold the data
```
Comment (howjmay) [inline connectivity/conn_http.h:77]: One extra redundant space after `Buffer`.

File: connectivity/conn_http.h [hunk 23]
```diff
@@ -23,31 +23,123 @@ extern "C" {
...
+ */
 retcode_t http_close(connect_info_t *const info);
 
+/**
+ * @brief Set post request message
+ *
+ * @param[in] api API for post request to https server, example: transaction/. Must be string.
+ * @param[in] host Http host to connect
+ * @param[in] port Http port to connect
+ * @param[in] req_body Pointer to post request body
```
Comment (howjmay) [inline connectivity/conn_http.h]: `Pointer of post request body`

**Review by howjmay — CHANGES_REQUESTED** [review]

File: connectivity/conn_http.h [hunk 24]
```diff
@@ -23,31 +23,123 @@ extern "C" {
...
+  mbedtls_ssl_config *ssl_config;     /**< mbedtls_ssl_config context                 */
+  mbedtls_x509_crt *cacert;           /**< mbedtls_x509 container                     */
 } connect_info_t;
 
+/**
+ * @brief Open http connection
+ *
+ * @param[in, out] info Context for http connection
+ * @param[in] seed_nonce Personalization data, that is device-specific identifiers. Can be NULL.
+ * @param[in] host Http host to connect
```
Comment (howjmay) [inline connectivity/conn_http.h]: How about using capitial `HTTP` instead of `Http`?

File: connectivity/conn_http.h (hunk 22 above)
Comment (splasky) [reply connectivity/conn_http.h:77]: There is only one space?

File: connectivity/conn_http.h (hunk 21 above)
Comment (jserv) [reply connectivity/conn_http.h]: I care about the grammar.

**Review by howjmay — CHANGES_REQUESTED** [review]

File: connectivity/conn_http.h [hunk 25]
```diff
@@ -23,31 +23,123 @@ extern "C" {
...
+ * - #RET_HTTP_CERT failed on http certificate setting error
+ * - #RET_HTTP_SSL failed on http ssl setting error
+ * - #RET_OK on success
+ * @see #retcode_t
+ */
 retcode_t http_open(connect_info_t *const info, char const *const seed_nonce, char const *const host,
                     char const *const port);
+
+/**
+ * @brief Send request to http connection
```
Comment (howjmay) [inline connectivity/conn_http.h]: Send request via HTTP protocol

File: connectivity/conn_http.h [hunk 26]
```diff
@@ -23,31 +23,123 @@ extern "C" {
...
+ *
+ * @param[in] info Context for http connection
+ * @param[in] req  Buffer holding the data
```
Comment (howjmay) [inline connectivity/conn_http.h]: redundant one space after `req`

File: connectivity/conn_http.h (hunk 22 above)
Comment (howjmay) [reply connectivity/conn_http.h:77]: Amend to `Buffer that will hold the response data`

File: connectivity/conn_http.h [hunk 27]
```diff
@@ -23,31 +23,123 @@ extern "C" {
...
+ *
+ * @return
+ * - #RET_OK on success
+ * - non-zero on error
+ * @see #retcode_t
+ */
 retcode_t http_close(connect_info_t *const info);
 
+/**
+ * @brief Set post request message
```
Comment (howjmay) [inline connectivity/conn_http.h]: Use capital capital `POST`

File: connectivity/conn_http.h [hunk 28]
```diff
@@ -23,31 +23,123 @@ extern "C" {
...
+ *
+ * @param[in] api API for post request to https server, example: transaction/. Must be string.
+ * @param[in] host Http host to connect
```
Comment (howjmay) [inline connectivity/conn_http.h]: use capital

File: connectivity/conn_http.h [hunk 29]
```diff
@@ -23,31 +23,123 @@ extern "C" {
...
+ * - #RET_OK on success
+ * - non-zero on error
+ * @see #retcode_t
+ */
 retcode_t http_close(connect_info_t *const info);
 
+/**
+ * @brief Set post request message
+ *
+ * @param[in] api API for post request to https server, example: transaction/. Must be string.
```
Comment (howjmay) [inline connectivity/conn_http.h]: The parameter `api` is somehow misleading. Could you name it as `path`

Comment (howjmay) [reply connectivity/conn_http.h]: And revise the desctiption as `API path for POST request to http(s) server, i.e. "transaction/". It must be in string`

File: connectivity/conn_http.h [hunk 30]
```diff
@@ -23,31 +23,123 @@ extern "C" {
...
+ * @param[in] host Http host to connect
+ * @param[in] port Http port to connect
```
Comment (howjmay) [inline connectivity/conn_http.h]: Assigning HTTP port

File: connectivity/conn_http.h [hunk 31]
```diff
@@ -23,31 +23,123 @@ extern "C" {
...
+ * @param[in] req_body Pointer of post request body
```
Comment (howjmay) [inline connectivity/conn_http.h]: POST

File: connectivity/conn_http.h [hunk 32]
```diff
@@ -23,31 +23,123 @@ extern "C" {
...
+ * @param[out] out Post request message
```
Comment (howjmay) [inline connectivity/conn_http.h]: POST

File: connectivity/conn_http.h [hunk 33]
```diff
@@ -23,31 +23,123 @@ extern "C" {
...
+ * @return
+ * - #RET_OK on success
+ * - #RET_OOM failed on out of memory error
+ * - non-zero on error
+ * @see #retcode_t
+ */
 retcode_t set_post_request(char const *const api, char const *const host, const uint32_t port,
                            char const *const req_body, char **out);
+/**
+ * @brief Set get request message
```
Comment (howjmay) [inline connectivity/conn_http.h]: GET

File: connectivity/conn_http.h [hunk 34]
```diff
@@ -23,31 +23,123 @@ extern "C" {
...
+ *
+ * @param[in] api API for post request to https server, example: transaction/. Must be string.
```
Comment (howjmay) [inline connectivity/conn_http.h]: GET request
Follow the format in `set_post_request`

File: connectivity/conn_http.h [hunk 35]
```diff
@@ -23,31 +23,123 @@ extern "C" {
...
+ * @param[in] host Http host to connect
```
Comment (howjmay) [inline connectivity/conn_http.h]: ditto

**Review by howjmay — CHANGES_REQUESTED** [review]

File: connectivity/conn_http.h [hunk 36]
```diff
@@ -23,31 +23,125 @@ extern "C" {
 #include "mbedtls/net.h"
 #include "mbedtls/ssl.h"
 
-#define HTTP_OK 200
-
+/**< Global pointer points to HTTP resource body for passing response body to http_parser.
```
Comment (jserv) [inline connectivity/conn_http.h]: Should be "A global pointer that ..."

File: connectivity/conn_http.h [hunk 37]
```diff
@@ -23,31 +23,125 @@ extern "C" {
...
+ * This variable will be removed or changed to static pointer as soon as possible. Don't
```
Comment (jserv) [inline connectivity/conn_http.h:27]: Use "FIXME" to denote.

File: utils/crypto_utils.h [hunk 38]
```diff
@@ -20,16 +20,94 @@ extern "C" {
...
+ * @param[in] plaintext The text to be encrypted
+ * @param[in] plaintext_len Plaintext length
+ * @param[out] ciphertext Ciphrtext
+ * @param[in] ciphertext_len Ciphertext length
+ * @param[in, out] iv[AES_BLOCK_SIZE] Initialization vector
+ * @param[in] key[AES_CBC_KEY_SIZE] Encryption key
+ * @param[in] device_id[IMSI_LEN + 1] Device id
+ *
+ * @return
+ * - Ciphertext text new length on success
```
Comment (jserv) [inline utils/crypto_utils.h:57]: Take care of "Ciphertext" for its representation.

File: utils/protocol.h [hunk 39]
```diff
@@ -12,12 +12,12 @@
...
 
 /**
- * @brief Send message via http protocol
+ * @brief Send message via HTTP protocol
  *
- * @param[in] host Https host
- * @param[in] port Https port
- * @param[in] api API for post request to https server, example: transaction/. Must be
- * string.
+ * @param[in] host HTTP host
```
Comment (jserv) [inline utils/protocol.h]: Https -> HTTP ?

File: utils/protocol.h [hunk 40]
```diff
@@ -12,12 +12,12 @@
...
+ * @param[in] port HTTP port
+ * @param[in] api API path for POST request to HTTP(S) server, i.e "transaction/". It
```
Comment (jserv) [inline utils/protocol.h]: Check indention.

**Pull request merged**